kb/development/script-coding-standards.md

```diff
@@ -0,0 +1,102 @@
+
+All scripts should start with a comment that explains how to run the script. Printing those explanations when the script is invoked with no argument or with `--help` is a plus.
+
+### File structure
+
+All executable scripts committed into the repository should have the following structure:
+
+* The first line is a [shebang line](https://en.wikipedia.org/wiki/Shebang_(Unix)), e.g. `#!/usr/bin/env python3`.
+* A brief description of the file (either as a comment or as a string).
+* Copyright notice and license indication.
+* Initialization code.
+* Auxiliary functions (unless the script is so simple it doesn't need any).
+* Code that performs the script's purpose (this can be as simple as calling a `main` function, or more complex).
+
+## Python coding standards
+
+### Directory structure
+
+Non-executable Python modules are in `scripts/mbedtls_dev`.
+Executable Python scripts in `scripts` can use auxiliary modules with `from mbedtls_dev import …`.
+
+Executable Python scripts in `tests/scripts` cannot refer to modules in `scripts/mbedtls_dev` directly, because we do not assume that `PYTHONPATH` contains the Mbed TLS source tree. To use other Python modules from a script in `tests/scripts`, use the following idiom:
+
+```python
+import scripts_path # pylint: disable=unused-import
+from mbedtls_dev import …
+```
+
+### Python imports
+
+We follow the usual convention that imports go at the top of a file (after the docstring and the license comment). Imports are typically ordered as follows:
+
+1. Standard library, in lexicographic order.
+2. Third-party libraries, if any, in lexicographic order.
+3. Modules from `mbedtls_dev`, in lexicographic order.
+
+Avoid importing individual names from modules, as it can be difficult to figure out which module an unqualified name comes from. It's ok to import names if they're sufficiently recognizable, for example:
+
+* Importing individual modules from `mbedtls_dev` is fine.
+* Importing capitalized type names from `typing` is fine.
+* Importing individual submodules from Cryptodome is fine.
+
+### Third-party Python libraries
+
+Avoid bringing in new dependencies, because each dependency adds risk to the project. Only use dependencies if they have a clear advantage.
+
+The minimum supported version of third-party Python libraries is recorded in `*.requirements.txt` files in the [`scripts` directory](https://github.com/Mbed-TLS/mbedtls/tree/development/scripts). As with the [Python version](#language-versions), keep in mind that long-time support branches may need to support older versions.
+
+### Python language features
+
+The maintainers of Mbed TLS are primarily C programmers, not Python programmers. It's ok to use fancy language features if they're useful, but don't necessarily assume that the reader is familiar with them. Make sure that if a reader is unfamiliar with an advanced feature, they'll be able to easily look it up.
+
+* Classes, methods and simple inheritance can be used freely.
```

Comment:
We currently support Python 3.6, so I don't think we can use this feature in even in development.

Reply:
Since the time we said we support Python 3.6, it's gone out of support. So we could switch to 3.7 as the minimum version in the development branch (and for nonessential backported scripts).

Data classes do save a minor annoyance (sometimes you just need a struct), and they aren't difficult to understand, so when they are permitted by language version considerations, I don't see why we should forbid them.